src/NativeBridge/UnixInterface.h

```diff
@@ -284,7 +289,7 @@ class UnixMlNetInterface
             // TRUSTED_PLATFORM_ASSEMBLIES
             tpaList.c_str(),
             // APP_PATHS
-            libsRoot,
+            dprepDirRoot,
```

Comment:
This seems wrong. Why is the DataPrep directory considered the `APP_PATHS`? Shouldn't it just instead be added to `tpaList`?

Reply:
DataPrep folder contains numerous dlls for azureml-dataprep package to run. Not all of them needed for in process integration that NimbusML doing here.  Also many of dlls are duplicates of dotnetcore2 package (.NET Core CLR runtime). Putting DataPrep dlls on TPAList will 
1.  make the list huuge. 
2.  will have  lot of duplicates like System.*.dlls. I can of course filter out them, but this will be an additional logic step. 
3. If there are duplicates no guarantee which of them will be used in runtime. 

I want to avoid this and use only dotnetcore2, ML.NET,  Microsoft.DataPrep.* and Microsoft.DPrep.* dlls - these are put into TPA list. If for some reason I missed any of dlls that DPrep needs I have set probing path to Dprep folder. 

---
In reply to: [302699171](https://github.com/microsoft/NimbusML/pull/181#discussion_r302699171) [](ancestors = 302699171)

Reply:
Why are we treating the ML.NET and the DataPrep libraries differently? Maybe we should just put  both locations on the `APP_PATHS`.

That way, when the next library comes along that we need to do this for, it is obvious to just add it to `APP_PATHS`.

Reply:
Yes, next library will go same way as Dprep into APP_PATHS. Reasons I treat ML.NET and DataPrep differently bcs:
1. Dprep is optional dependency, if its not installed NimbusML should still work in all scenario except taking input from Dprep files. ML.NET is a mandatory core dependency here. I am also very familiar with all the needed dlls. So felt TPAList is more appropriate here to list all of them.   
2. Want to be on a safer side. I am not familiar how probing with APP_PATHS work. If there are duplicate dlls in ML.NET and Dprep folder, will first in wins? With TPAList this seems not guaranteed, so could be same with APP-PATHS. I want to ensure NimbusML core ML.NET uses are untouched by this so used TPA List. 
3. With Dprep dlls - there are seems tons of them, which ones are necessary for my case and which ones are not, difficult to figure. So I packaged only few ones (<3MB total size) that are needed for JIT compilation when running NimbusML and set the APP_PATHS  to Dprep folder for the rest.
Another issue there are mismatches in versions of *Dprep.dlls that are exposed in NuGet  and installed with azureml-dataprep package. During testing I found out that I need to keep the versions of *Dprep.dll that I built against but for the rest of supporting dlls I can point to Dprep folder. We will have to do further testing to figure out correct versions of azureml-dataprep to be installed to work with built ones in NimbusML. Now it seems that latest azureml-dataprep package works.

---
In reply to: [302984801](https://github.com/microsoft/NimbusML/pull/181#discussion_r302984801) [](ancestors = 302984801)

Reply:
After reading https://docs.microsoft.com/en-us/dotnet/core/tutorials/netcore-hosting#step-3---prepare-runtime-properties, and talking with a CoreCLR dev, the recommendation here is to use the `TRUSTED_PLATFORM_ASSEMBLIES` for all the assemblies. From the doc:

> Because the host has more control over which assemblies are loaded using the TPA list, it is a best practice for hosts to determine which assemblies they expect to load and list them explicitly.

I don't want to block you going this route, since I haven't really worked in NimbusML, but I just figured I'd give you as much information as possible to make an informed decision.

Reply:
thank you Eric!

---
In reply to: [303107371](https://github.com/microsoft/NimbusML/pull/181#discussion_r303107371) [](ancestors = 303107371)